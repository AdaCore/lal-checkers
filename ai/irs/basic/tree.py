"""
Provides the tree constructors of the Basic intermediate representation.
"""

from ai.utils import Bunch


def _visitable(name):
    """
    A class decorator that adds to a node class an entry point for visitors of
    this node. The given method name will be called on the visitor with the
    visited node.

    :param str name: The name of the method used to visit this type of node.

    :return: A function which can be called on a class to augment it with
        visiting capabilities.

    :rtype: type -> type
    """
    def enter_visitor(node, visitor, *args):
        """
        :param Node node: The visited node.
        :param visitors.Visitor visitor: The node visitor.
        :param *object args: Additional arguments for the visitor.
        :return: Whatever the visitor returns.
        :rtype: object
        """
        return getattr(visitor, name)(node, *args)

    def wrapper(cls):
        """
        :param type cls: The class being augmented with visiting capabilities.

        :return: The same class, with an additional method allowing visitors
            to visit its instances.

        :rtype: type
        """
        cls.visit = enter_visitor
        return cls

    return wrapper


class Node(object):
    """
    The base class for any Basic tree node.
    """
    def __init__(self, **data):
        """
        Initializes the node with the given user data.

        :param **object data: User-defined data.
        """
        self.data = Bunch(**data)

    def children(self):
        """
        :return: The children of this node.
        :rtype: iterable[Node]
        """
        raise NotImplementedError


@_visitable("visit_program")
class Program(Node):
    """
    The node which represents a "program".

    It consists of a list of statements that are to be executed one after
    the other.
    """
    def __init__(self, stmts, **data):
        """
        :param list[Stmt] stmts: The list of statements this program consists
            of.

        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.stmts = stmts

    def children(self):
        for stmt in self.stmts:
            yield stmt


@_visitable("visit_ident")
class Identifier(Node):
    """
    An identifier, like "x", etc.
    """
    def __init__(self, var, **data):
        """
        :param Variable var: The variable object associated to this identifier.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.var = var

    def children(self):
        yield self.var


@_visitable("visit_var")
class Variable(Node):
    """
    Represents a unique variable object, that can be referred to by multiple
    identifiers.
    """
    def __init__(self, name, **data):
        """
        :param string name: The name of the variable
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.name = name

    def children(self):
        return iter(())


class Stmt(Node):
    """
    The base class for all statements.
    """
    pass


@_visitable("visit_assign")
class AssignStmt(Stmt):
    """
    Represents the assign statement, i.e. [identifier] = [expr].
    """
    def __init__(self, id, expr, **data):
        """
        :param Identifier id: The identifier being assigned.
        :param Expr expr: The expression assigned to the identifier.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.id = id
        self.expr = expr

    def children(self):
        yield self.id
        yield self.expr


@_visitable("visit_split")
class SplitStmt(Stmt):
    """
    A control-flow statement representing a nondeterministic choice of
    execution path, i.e. the two branches are visited.
    """
    def __init__(self, branches, **data):
        """
        :param list[list[Stmt]] branches: The branches of the split statement,
            where each branch is a list of statements. At least two branches
            are expected.

        :param **object data: User-defined data.
        """
        assert len(branches) >= 2

        Node.__init__(self, **data)
        self.branches = branches

    def children(self):
        for branch in self.branches:
            for stmt in branch:
                yield stmt


@_visitable("visit_loop")
class LoopStmt(Stmt):
    """
    Represents a nondeterministic loop.
    """
    def __init__(self, stmts, **data):
        """
        :param list[Stmt] stmts: The list of statements appearing in the body
            of this loop.

        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.stmts = stmts

    def children(self):
        for stmt in self.stmts:
            yield stmt


@_visitable("visit_label")
class LabelStmt(Stmt):
    """
    Represents a point in the program which can be jumped to.
    """
    def __init__(self, name, **data):
        """
        :param str name: The name of the label.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.name = name

    def children(self):
        return iter(())


@_visitable("visit_goto")
class GotoStmt(Stmt):
    """
    Represents a jump statement.
    """
    def __init__(self, label, **data):
        """
        :param LabelStmt label: The label to jump to.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.label = label

    def children(self):
        yield self.label


@_visitable("visit_read")
class ReadStmt(Stmt):
    """
    Represents the havoc operation on a variable.
    """
    def __init__(self, id, **data):
        """
        :param Identifier id: The identifier being read.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.id = id

    def children(self):
        yield self.id


@_visitable("visit_use")
class UseStmt(Stmt):
    """
    Represents the fact that a variable was used at this point.
    """
    def __init__(self, id, **data):
        """
        :param Identifier id: The identifier being used.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.id = id

    def children(self):
        yield self.id


@_visitable("visit_assume")
class AssumeStmt(Stmt):
    """
    Represents the fact that an expression is assumed to be true at this point.
    """
    def __init__(self, expr, **data):
        """
        :param Expr expr: The expression being assumed.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.expr = expr

    def children(self):
        yield self.expr


class Expr(Node):
    """
    Base class for expression nodes.
    """
    pass


@_visitable("visit_funcall")
class FunCall(Expr):
    """
    Represents a function call.
    """
    def __init__(self, fun_id, args, **data):
        """
        :param object fun_id: An object identifying the function called.
        :param list[Expr] args: The arguments passed.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.fun_id = fun_id
        self.args = args

    def children(self):
        for arg in self.args:
            yield arg


@_visitable("visit_lit")
class Lit(Expr):
    """
    Represents a literal value.
    """
    def __init__(self, val, **data):
        """
        :param object val: Any value.
        :param **object data: User-defined data.
        """
        Node.__init__(self, **data)
        self.val = val

    def children(self):
        return iter(())
