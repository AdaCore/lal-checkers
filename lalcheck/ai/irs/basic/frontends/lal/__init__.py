"""
Provides a libadalang frontend for the Basic IR.
"""

import libadalang as lal
from lalcheck.ai.utils import profile
from lalcheck.tools.logger import log_stdout, log

import analysis
import typers
import utils
import time
from codegen import ConvertUniversalTypes, gen_ir


unknown_typer = typers.unknown_typer


class ExtractionContext(object):
    """
    The libadalang-based frontend interface. Provides method for extracting
    IR programs from Ada source files (see extract_programs), as well as
    a default typer for those programs (see default_typer).

    Note: programs extracted using different ExtractionContext are not
    compatible. Also, this extraction context must be kept alive as long
    as the programs parsed with it are intended to be used.
    """
    def __init__(self, lal_ctx=None):
        self.lal_ctx = lal.AnalysisContext() if lal_ctx is None else lal_ctx

        # Get a dummy node, needed to call static properties of libadalang.
        dummy = self.lal_ctx.get_from_buffer(
            "<dummy>", 'package Dummy is end;'
        ).root

        # Find the Character TypeDecl.
        char_type = dummy.p_standard_unit.root.find(
            lambda x: x.is_a(lal.TypeDecl) and x.f_name.text == "Character"
        )

        float_type = dummy.p_standard_unit.root.find(
            lambda x: x.is_a(lal.TypeDecl) and x.f_name.text == "Float"
        )

        self.evaluator = utils.ConstExprEvaluator(
            dummy.p_bool_type,
            dummy.p_int_type,
            float_type,
            char_type,
            dummy.p_universal_int_type,
            dummy.p_universal_real_type
        )

        self.type_models = {}
        self.fun_models = {}
        self.subpdata = {}
        self._internal_typer = self.default_typer()

    @staticmethod
    def for_project(project_file, scenario_vars={}):
        return ExtractionContext(lal.AnalysisContext(
            unit_provider=lal.UnitProvider.for_project(
                project_file,
                scenario_vars
            )
        ))

    @staticmethod
    def empty():
        return ExtractionContext()

    def extract_programs_from_file(self, ada_file):
        """
        :param str ada_file: A path to the Ada source file from which to
            extract programs.

        :return: a Basic IR Program for each subprogram body that exists in the
            given source code.

        :rtype: iterable[irt.Program]
        """
        return self._extract_from_unit(self.lal_ctx.get_from_file(ada_file))

    def extract_programs_from_provider(self, name, kind):
        return self._extract_from_unit(self.lal_ctx.get_from_provider(
            name, kind
        ))

    def extract_programs_from_unit(self, unit):
        """
        :param lal.AnalysisUnit unit: The already parsed compilation unit.
        """
        return self._extract_from_unit(unit)

    @profile()
    def use_model(self, name):
        model_unit = self.lal_ctx.get_from_provider(name, "specification")
        with log_stdout('error'):
            for diag in model_unit.diagnostics:
                print('   {}'.format(diag))
                return

        model_ofs = model_unit.root.findall(
            lambda x: (x.is_a(lal.AspectAssoc)
                       and x.f_id.text.lower() == "model_of")
        )

        type_models = {
            aspect.parent.parent.parent: aspect.f_expr.p_referenced_decl
            for aspect in model_ofs
            if aspect.parent.parent.parent.is_a(lal.TypeDecl, lal.SubtypeDecl)
        }

        fun_models = {
            aspect.parent.parent.parent: aspect.f_expr.p_referenced_decl
            for aspect in model_ofs
            if aspect.parent.parent.parent.is_a(lal.SubpDecl)
        }

        if len(type_models) + len(fun_models) != len(model_ofs):
            with log_stdout('info'):
                print("warning: detected usage of Model_Of in an unknown "
                      "context.")

        for tdecl, ref in type_models.iteritems():
            if ref is None:
                with log_stdout('info'):
                    print("warning: Model_Of used on '{}' refers to an unknown"
                          " type.".format(tdecl.f_name.text))
            else:
                self.type_models[ref] = tdecl

        for fdecl, ref in fun_models.iteritems():
            if ref is None:
                with log_stdout('info'):
                    print(
                        "warning: Model_Of used on '{}' refers to an unknown "
                        "procedure/function.".format(
                            fdecl.f_subp_spec.f_subp_name.text
                        )
                    )
            else:
                self.fun_models[ref] = fdecl

    @profile()
    def _extract_from_unit(self, unit):
        if unit.root is None:
            with log_stdout('error'):
                print('Could not parse {}:'.format(unit.filename))
                for diag in unit.diagnostics:
                    print('   {}'.format(diag))
                    return

        unit.populate_lexical_env()

        subpdata = analysis.traverse_unit(unit.root)
        self.subpdata.update(subpdata)

        progs = []
        for subp, subpuserdata in subpdata.iteritems():
            if subp.is_a(lal.BaseSubpBody):
                start_t = time.clock()
                progs.append(
                    gen_ir(self, subp, self._internal_typer, subpuserdata)
                )
                end_t = time.clock()
                log(
                    'timings',
                    " - Transformation of subprocedure {} took {}s".format(
                        subp.f_subp_spec.f_subp_name.text,
                        end_t - start_t
                    )
                )

        converter = ConvertUniversalTypes(self.evaluator, self._internal_typer)

        for prog in progs:
            prog.visit(converter)

        return progs

    def standard_typer(self):
        """
        :return: A Typer for Ada standard types of programs parsed using
            this extraction context.

        :rtype: types.Typer[lal.AdaNode]
        """
        return typers.standard_typer(self)

    def model_typer(self, inner_typer):
        return typers.model_typer(self, inner_typer)

    def default_typer(self, fallback_typer=None):
        """
        :return: The default Typer for Ada programs parsed using this
            extraction context.

        :rtype: types.Typer[lal.AdaNode]
        """
        return typers.default_typer(self, fallback_typer)
