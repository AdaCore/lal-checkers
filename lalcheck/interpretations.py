"""
Defines the TypeInterpreter interface, as well as a few common
TypeInterpreters.
"""

from lalcheck.domain_ops import (
    boolean_ops,
    interval_ops,
    finite_lattice_ops
)
from lalcheck.constants import ops, lits
from lalcheck import types
from lalcheck import domains


class TypeInterpreter(object):
    """
    A TypeInterpreter is used to determine how a type is interpreted in the
    backend. It must provide the domain used, a set of definitions that are
    available on this new domain, as well as a builder function to build
    elements of this domain from literal values.
    """
    def from_type(self, tpe):
        """
        Given a type, returns the domain that this TypeInterpreter would use to
        represent the type, a dictionary of definitions, as well as a builder
        for this domain. Returns None if this interpreter does not provide any
        interpretation for the given type.
        """
        raise NotImplementedError

    def __or__(self, other):
        """
        Creates a new interpreter by combining two interpreters. This new
        TypeInterpreter will try to interpret the given type using the first
        TypeInterpreter. It it failed (returned None), it will interpret it
        using the second TypeInterpreter.
        """
        @type_interpreter
        def f(hint):
            x = self.from_type(hint)
            return x if x is not None else other.from_type(hint)

        return f


def type_interpreter(fun):
    """
    A useful decorator to use on a function to turn it into an TypeInterpreter.
    The decorated function must receive a type as parameter and return
    a type interpretation.
    """
    class AnonymousInterpreter(TypeInterpreter):
        def from_type(self, tpe):
            return fun(tpe)

    return AnonymousInterpreter()


def dict_to_provider(def_dict):
    """
    Converts a dictionnary of definitions indexed by their names and domain
    signatures to a provider function.
    """
    def provider(name, signature):
        if (name, signature) in def_dict:
            return def_dict[name, signature]

    return provider


@type_interpreter
def default_boolean_interpreter(tpe):
    if tpe.is_a(types.Boolean):
        bool_dom = boolean_ops.Boolean
        un_fun_dom = (bool_dom, bool_dom)
        bin_fun_dom = (bool_dom, bool_dom, bool_dom)

        defs = {
            (ops.Not, un_fun_dom): boolean_ops.not_,
            (ops.And, bin_fun_dom): boolean_ops.and_,
            (ops.Or, bin_fun_dom): boolean_ops.or_,

            (ops.Eq, bin_fun_dom): finite_lattice_ops.eq(bool_dom),
            (ops.Neq, bin_fun_dom): finite_lattice_ops.neq(bool_dom)
        }

        inv_defs = {
            (ops.Not, un_fun_dom): boolean_ops.inv_not,
            (ops.And, bin_fun_dom): boolean_ops.inv_and,
            (ops.Or, bin_fun_dom): boolean_ops.inv_or,

            (ops.Eq, bin_fun_dom): finite_lattice_ops.inv_eq(bool_dom),
            (ops.Neq, bin_fun_dom): finite_lattice_ops.inv_neq(bool_dom)
        }

        builder = boolean_ops.lit

        return (
            bool_dom,
            dict_to_provider(defs),
            dict_to_provider(inv_defs),
            builder
        )


@type_interpreter
def default_int_range_interpreter(tpe):
    if tpe.is_a(types.IntRange):
        int_dom = domains.Intervals(tpe.frm, tpe.to)
        bool_dom = boolean_ops.Boolean
        unary_fun_dom = (int_dom, int_dom)
        binary_fun_dom = (int_dom, int_dom, int_dom)
        binary_rel_dom = (int_dom, int_dom, bool_dom)

        defs = {
            (ops.Plus, binary_fun_dom):
                interval_ops.add_no_wraparound(int_dom),

            (ops.Minus, binary_fun_dom):
                interval_ops.sub_no_wraparound(int_dom),

            (ops.Lt, binary_rel_dom): interval_ops.lt(int_dom),
            (ops.Le, binary_rel_dom): interval_ops.le(int_dom),
            (ops.Eq, binary_rel_dom): interval_ops.eq(int_dom),
            (ops.Neq, binary_rel_dom): interval_ops.neq(int_dom),
            (ops.Ge, binary_rel_dom): interval_ops.ge(int_dom),
            (ops.Gt, binary_rel_dom): interval_ops.gt(int_dom),
            (ops.Neg, unary_fun_dom): interval_ops.inverse(int_dom),
        }

        inv_defs = {
            (ops.Plus, binary_fun_dom):
                interval_ops.inv_add_no_wraparound(int_dom),

            (ops.Minus, binary_fun_dom):
                interval_ops.inv_sub_no_wraparound(int_dom),

            (ops.Lt, binary_rel_dom): interval_ops.inv_lt(int_dom),
            (ops.Le, binary_rel_dom): interval_ops.inv_le(int_dom),
            (ops.Eq, binary_rel_dom): interval_ops.inv_eq(int_dom),
            (ops.Neq, binary_rel_dom): interval_ops.inv_neq(int_dom),
            (ops.Ge, binary_rel_dom): interval_ops.inv_ge(int_dom),
            (ops.Gt, binary_rel_dom): interval_ops.inv_gt(int_dom),
            (ops.Neg, unary_fun_dom): interval_ops.inv_inverse(int_dom)
        }

        builder = interval_ops.lit(int_dom)

        return (
            int_dom,
            dict_to_provider(defs),
            dict_to_provider(inv_defs),
            builder
        )


@type_interpreter
def default_enum_interpreter(tpe):
    if tpe.is_a(types.Enum):
        enum_dom = domains.FiniteLattice.of_subsets(set(tpe.lits))
        bool_dom = boolean_ops.Boolean
        bin_rel_dom = (enum_dom, enum_dom, bool_dom)

        defs = {
            (ops.Eq, bin_rel_dom): finite_lattice_ops.eq(enum_dom),
            (ops.Neq, bin_rel_dom): finite_lattice_ops.neq(enum_dom)
        }

        inv_defs = {
            (ops.Eq, bin_rel_dom): finite_lattice_ops.inv_eq(enum_dom),
            (ops.Neq, bin_rel_dom): finite_lattice_ops.inv_neq(enum_dom)
        }

        builder = finite_lattice_ops.lit(enum_dom)

        return (
            enum_dom,
            dict_to_provider(defs),
            dict_to_provider(inv_defs),
            builder
        )


@type_interpreter
def simple_access_interpreter(tpe):
    if tpe.is_a(types.Pointer):
        ptr_dom = domains.FiniteLattice.of_subsets({lits.Null, lits.NotNull})
        bool_dom = boolean_ops.Boolean
        bin_rel_dom = (ptr_dom, ptr_dom, bool_dom)

        defs = {
            (ops.Eq, bin_rel_dom): finite_lattice_ops.eq(ptr_dom),
            (ops.Neq, bin_rel_dom): finite_lattice_ops.neq(ptr_dom)
        }

        inv_defs = {
            (ops.Eq, bin_rel_dom): finite_lattice_ops.inv_eq(ptr_dom),
            (ops.Neq, bin_rel_dom): finite_lattice_ops.inv_neq(ptr_dom)
        }

        builder = finite_lattice_ops.lit(ptr_dom)
        null = builder(lits.Null)
        notnull = builder(lits.NotNull)

        def def_provider(name, sig):
            if (name, sig) in defs:
                return defs[name, sig]

            elif (name == ops.Deref and
                    len(sig) == 2 and
                    sig[0] == ptr_dom):
                elem_dom = sig[1]

                def deref(ptr):
                    return elem_dom.bottom if ptr == null else elem_dom.top

                return deref

            elif (name == ops.Address and
                    len(sig) == 2 and
                    sig[1] == ptr_dom):
                elem_dom = sig[0]

                def address(elem):
                    return notnull

                return address

        def inv_def_provider(name, sig):
            if (name, sig) in defs:
                return inv_defs[name, sig]

            elif (name == ops.Deref and
                    len(sig) == 2 and
                    sig[0] == ptr_dom):
                elem_dom = sig[1]

                def inv_deref(elem, e_constr):
                    if ptr_dom.is_empty(e_constr) or elem_dom.is_empty(elem):
                        return None

                    if ptr_dom.le(notnull, e_constr):
                        return notnull

                    return None

                return inv_deref

            elif (name == ops.Address and
                    len(sig) == 2 and
                    sig[1] == ptr_dom):
                elem_dom = sig[0]

                def inv_address(ptr, e_constr):
                    if ptr_dom.is_empty(ptr) or elem_dom.is_empty(e_constr):
                        return None

                    return e_constr

                return inv_address

        return (
            ptr_dom,
            def_provider,
            inv_def_provider,
            builder
        )


default_type_interpreter = (
    default_boolean_interpreter |
    default_int_range_interpreter |
    default_enum_interpreter |
    simple_access_interpreter
)
