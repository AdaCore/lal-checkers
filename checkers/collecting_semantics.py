"""
An analyzer that collects semantics at each program point.
"""

from lalcheck.irs.basic.tools import (
    CFGBuilder,
    ExprEvaluator,
    ExprSolver
)

from lalcheck.irs.basic.tree import Variable
from lalcheck.irs.basic.tools import PrettyPrinter
from lalcheck.irs.basic.purpose import SyntheticVariable
from lalcheck.irs.basic import visitors
from lalcheck.utils import KeyCounter
from lalcheck.digraph import Digraph
from lalcheck import domains
from lalcheck import dot_printer

from xml.sax.saxutils import escape
from collections import defaultdict


class VarTracker(visitors.CFGNodeVisitor):
    def __init__(self, var_set, vars_domain, vars_idx, evaluator, c_solver):
        self.vars = var_set
        self.vars_domain = vars_domain
        self.vars_idx = vars_idx
        self.evaluator = evaluator
        self.constr_solver = c_solver

    def state_to_env(self, state):
        return {v: state[self.vars_idx[v]] for v in self.vars}

    def env_to_state(self, env):
        if len(env) == 0:
            return self.vars_domain.bottom
        else:
            return tuple(env[v] for v in self.vars)

    def visit_assign(self, assign, state):
        env = self.state_to_env(state)
        env[assign.id.var] = self.evaluator.eval(assign.expr, env)
        return self.env_to_state(env)

    def visit_assume(self, assume, state):
        return self.env_to_state(self.constr_solver.solve(
            assume.expr,
            self.state_to_env(state)
        ))

    def visit_read(self, read, state):
        return tuple(self.vars_domain.top[i] if i == self.vars_idx[read.id.var]
                     else x for i, x in enumerate(state))

    def visit_use(self, use, state):
        return state


class SimpleTraceLattice(domains.FiniteSubsetLattice):
    def __init__(self, *args):
        super(SimpleTraceLattice, self).__init__(*args)

    def update(self, a, b, widen=False):
        return super(SimpleTraceLattice, self).update(a, b, False)


class MergePredicateBuilder(object):
    def __init__(self, predicate):
        self.predicate = predicate

    def __or__(self, other):
        def f(trace_domain, vals_domain):
            s_pred = self.build(trace_domain, vals_domain)
            o_pred = other.build(trace_domain, vals_domain)

            return lambda a, b: s_pred(a, b) or o_pred(a, b)

        return MergePredicateBuilder(f)

    def __and__(self, other):
        def f(trace_domain, vals_domain):
            s_pred = self.build(trace_domain, vals_domain)
            o_pred = other.build(trace_domain, vals_domain)

            return lambda a, b: s_pred(a, b) and o_pred(a, b)

        return MergePredicateBuilder(f)

    def build(self, trace_domain, vals_domain):
        return self.predicate(trace_domain, vals_domain)


def _mp_always(*_):
    return lambda *_: True


def _mp_never(*_):
    return lambda *_: False


def _mp_le_traces(trace_domain, _):
    return lambda a, b: trace_domain.le(a[0], b[0])


def _mp_eq_vals(_, vals_domain):
    return lambda a, b: vals_domain.eq(a[1], b[1])


MergePredicateBuilder.Always = MergePredicateBuilder(_mp_always)
MergePredicateBuilder.Never = MergePredicateBuilder(_mp_never)
MergePredicateBuilder.Le_Traces = MergePredicateBuilder(_mp_le_traces)
MergePredicateBuilder.Eq_Vals = MergePredicateBuilder(_mp_eq_vals)


def html_render_node(node):
    return escape(PrettyPrinter.pretty_print(node))


def save_cfg_to(file_name, cfg):
    def render_node(node):
        return (html_render_node(node),) if node is not None else ()

    def render_widening_point(is_widening_point):
        return (escape('<widening_point>'),) if is_widening_point else ()

    with open(file_name, 'w') as f:
        f.write(dot_printer.gen_dot(cfg, [
            dot_printer.DataPrinter('node', render_node),
            dot_printer.DataPrinter('is_widening_point', render_widening_point)
        ]))


def repr_trace(trace):
    return tuple(p.name for p in trace)


def build_resulting_graph(file_name, cfg, results, trace_domain):
    paths = defaultdict(list)

    var_set = {
        v
        for node, state in results.iteritems()
        for trace, values in state.iteritems()
        for v, _ in values.iteritems()
    }

    for node, state in results.iteritems():
        for trace, values in state.iteritems():
            paths[frozenset(trace)].append(
                Digraph.Node(
                    node.name,
                    ___orig=node,
                    **{
                        v.name: value
                        for v, value in values.iteritems()
                        if not SyntheticVariable.is_purpose_of(v)
                    }
                )
            )

    edges = []
    for trace, nodes in paths.iteritems():
        for node in nodes:
            predecessors = [
                n
                for t, ns in paths.iteritems()
                for n in ns
                if trace_domain.le(t, trace)
                if n.data.___orig in cfg.ancestors(node.data.___orig)
            ]

            for pred in predecessors:
                edges.append(Digraph.Edge(pred, node))

    res_graph = Digraph(
        [n for _, nodes in paths.iteritems() for n in nodes],
        edges
    )

    def print_orig(orig):
        if orig.data.node is not None:
            return (
                '<i>{}</i>'.format(html_render_node(orig.data.node)),
            )
        return ()

    def print_result_builder(name):
        return lambda value: (name, str(value))

    with open(file_name, 'w') as f:
        f.write(dot_printer.gen_dot(res_graph, [
            dot_printer.DataPrinter('___orig', print_orig)
        ] + [
            dot_printer.DataPrinter(v.name, print_result_builder(
                "{} &isin;".format(v.name)
            ))
            for v in var_set
        ]))


class AnalysisResults(object):
    """
    Contains the results of the collecting semantics analysis.
    """
    def __init__(self, cfg, semantics, trace_domain, vars_domain,
                 evaluator):
        self.cfg = cfg
        self.semantics = semantics
        self.trace_domain = trace_domain
        self.vars_domain = vars_domain
        self.evaluator = evaluator

    def save_cfg_to_file(self, file_name):
        """
        Prints the control-flow graph as a DOT file to the given file name.
        """
        save_cfg_to(file_name, self.cfg)

    def save_results_to_file(self, file_name):
        """
        Prints the resulting graph as a DOT file to the given file name.
        Displays the state of each variable at each program point.
        """
        build_resulting_graph(
            file_name,
            self.cfg,
            self.semantics,
            self.trace_domain
        )

    def eval_at(self, node, expr):
        """
        Given a program point, evaluates for each program trace available at
        this program point the given expression.
        """
        return {
            trace: self.evaluator.eval(expr, env)
            for trace, env in self.semantics[node].iteritems()
        }


def collect_semantics(prog, model, merge_pred_builder):

    cfg = prog.visit(CFGBuilder())

    var_set = set(visitors.findall(prog, lambda n: isinstance(n, Variable)))
    vars_idx = {v: i for i, v in enumerate(var_set)}
    vars_domain = domains.Product(*(model[v].domain for v in var_set))
    trace_domain = SimpleTraceLattice(cfg.nodes)

    evaluator = ExprEvaluator(model)
    solver = ExprSolver(model)

    widening_counter = KeyCounter()
    widening_delay = 10

    def do_widen(counter):
        # will widen when counter == widen_delay, then narrow
        return counter == widening_delay

    do_stmt = VarTracker(var_set, vars_domain, vars_idx, evaluator, solver)

    lat = domains.Set(
        domains.Product(
            trace_domain,
            vars_domain
        ),
        merge_pred_builder.build(
            trace_domain,
            vars_domain
        ),
        None  # We don't need a top element here.
    )

    def it(states):
        new_states = states.copy()

        for node in cfg.nodes:
            inputs = [new_states[anc] for anc in cfg.ancestors(node)]
            res = (lat.build([(trace_domain.bottom, vars_domain.top)])
                   if len(inputs) == 0 else reduce(lat.join, inputs))

            res = lat.build([
                (
                    trace_domain.join(trace, trace_domain.build([node])),
                    updated_val
                )
                for trace, values in res
                for updated_val in [
                    node.data.node.visit(do_stmt, values)
                    if node.data.node is not None else values
                ]
                if not vars_domain.is_empty(updated_val)
            ])

            if node.data.is_widening_point:
                if do_widen(widening_counter.get_incr(node)):
                    res = lat.update(new_states[node], res, True)

            new_states[node] = res

        return new_states

    result = {n: lat.bottom for n in cfg.nodes}
    last = None
    while result != last:
        last, result = result, it(result)

    formatted_results = {
        node: {
            trace: {
                v: values[vars_idx[v]] for v in var_set
            }
            for trace, values in state
        }
        for node, state in result.iteritems()
    }

    return AnalysisResults(
        cfg,
        formatted_results,
        trace_domain,
        vars_domain,
        evaluator
    )
