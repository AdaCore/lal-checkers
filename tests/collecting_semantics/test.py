import lalcheck.irs.basic.frontends.lal as lal2basic
from lalcheck.irs.basic.tools import Models
from lalcheck.interpretations import default_type_interpreter
from checkers.collecting_semantics import (
    collect_semantics,
    MergePredicateBuilder
)


def do(test_name, merge_predicate, idx=None):
    ctx = lal2basic.new_context()

    progs = lal2basic.extract_programs(
        ctx,
        '{}.adb'.format(test_name)
    )

    model_builder = Models(
        lal2basic.default_typer(ctx),
        default_type_interpreter
    )

    model = model_builder.of(*progs)

    collect_semantics(
        progs[0],
        model,
        merge_predicate,
        'output/{}_cfg.dot'.format(test_name),
        'output/{}{}_res.dot'.format(test_name, '' if idx is None else idx)
    )


le_trace_eq_vals = (
    MergePredicateBuilder.Le_Traces |
    MergePredicateBuilder.Eq_Vals
)

do('test_simple_if', MergePredicateBuilder.Always, '_always')
do('test_simple_if', le_trace_eq_vals, '_le_t_eq_v')

do('test_simple_if_expr', MergePredicateBuilder.Always, '_always')
do('test_simple_if_expr', le_trace_eq_vals, '_le_t_eq_v')

do('test_simple_while', le_trace_eq_vals)

do('test_simple_enum', MergePredicateBuilder.Always, '_always')

do('test_ptr', MergePredicateBuilder.Always, '_always')
do('test_ptr', le_trace_eq_vals, '_le_t_eq_v')
