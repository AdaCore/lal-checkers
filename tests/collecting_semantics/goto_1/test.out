{
  "always": {
    "assign1": [
      {
        "trace:": [
          "assign1",
          "read0",
          "start0"
        ],
        "values": {
          "x": "[1, 1]"
        }
      }
    ],
    "read0": [
      {
        "trace:": [
          "read0",
          "start0"
        ],
        "values": {
          "x": "[-2147483648, 2147483647]"
        }
      }
    ],
    "start0": [
      {
        "trace:": [
          "start0"
        ],
        "values": {
          "x": "[-2147483648, 2147483647]"
        }
      }
    ]
  },
  "le_t_eq_v": {
    "assign1": [
      {
        "trace:": [
          "assign1",
          "read0",
          "start0"
        ],
        "values": {
          "x": "[1, 1]"
        }
      }
    ],
    "read0": [
      {
        "trace:": [
          "read0",
          "start0"
        ],
        "values": {
          "x": "[-2147483648, 2147483647]"
        }
      }
    ],
    "start0": [
      {
        "trace:": [
          "start0"
        ],
        "values": {
          "x": "[-2147483648, 2147483647]"
        }
      }
    ]
  }
}